gtsfm/utils/metrics.py

```diff
@@ -55,21 +62,152 @@ def count_correct_correspondences(
+    if None not in [gt_scene_mesh, gt_wTi1, gt_wTi2]:
+        gt_camera_i1 = PinholeCameraCal3Bundler(gt_wTi1, intrinsics_i1)
+        gt_camera_i2 = PinholeCameraCal3Bundler(gt_wTi2, intrinsics_i2)
+        is_inlier, reproj_error = mesh_inlier_correspondences(
+            matched_keypoints_i1,
+            matched_keypoints_i2,
+            gt_camera_i1,
+            gt_camera_i2,
+            gt_scene_mesh,
+            dist_threshold,
+        )
+    elif gt_wTi1 is not None and gt_wTi2 is not None:
+        gt_i2Ti1 = gt_wTi2.between(gt_wTi1)
+        is_inlier, reproj_error = epipolar_inlier_correspondences(
+            matched_keypoints_i1,
+            matched_keypoints_i2,
+            intrinsics_i1,
+            intrinsics_i2,
+            gt_i2Ti1,
+            dist_threshold,
+        )
+    else:
+        return None, None
+
+    return is_inlier, reproj_error
+
+
+def epipolar_inlier_correspondences(
+    keypoints_i1: Keypoints,
+    keypoints_i2: Keypoints,
+    intrinsics_i1: Cal3Bundler,
+    intrinsics_i2: Cal3Bundler,
+    i2Ti1: Pose3,
+    dist_threshold: float,
+) -> Tuple[Optional[np.ndarray], Optional[np.ndarray]]:
+    """Compute inlier correspondences using epipolar geometry and the ground truth relative pose.
 
-    if len(keypoints_i1) == 0:
-        return 0
+    Args:
+        keypoints_i1: keypoints in image i1.
+        keypoints_i2: corr. keypoints in image i2.
+        intrinsics_i1: intrinsics for i1.
+        intrinsics_i2: intrinsics for i2.
+        i2Ti1: relative pose
+        dist_threshold: max acceptable distance for a correct correspondence.
 
+    Returns:
+        is_inlier: (N, ) mask of inlier correspondences.
+        distance_squared: squared sampson distance between corresponding keypoints.
+    """
     i2Ei1 = EssentialMatrix(i2Ti1.rotation(), Unit3(i2Ti1.translation()))
     i2Fi1 = verification_utils.essential_to_fundamental_matrix(i2Ei1, intrinsics_i1, intrinsics_i2)
-
     distance_squared = verification_utils.compute_epipolar_distances_sq_sampson(
         keypoints_i1.coordinates, keypoints_i2.coordinates, i2Fi1
     )
-    inlier_mask_gt = distance_squared < epipolar_dist_threshold ** 2
-    return inlier_mask_gt
+    is_inlier = distance_squared < dist_threshold ** 2 if distance_squared is not None else None
+
+    return is_inlier, distance_squared
+
+
+def mesh_inlier_correspondences(
+    keypoints_i1: Keypoints,
+    keypoints_i2: Keypoints,
+    gt_camera_i1: PinholeCameraCal3Bundler,
+    gt_camera_i2: PinholeCameraCal3Bundler,
+    gt_scene_mesh: trimesh.Trimesh,
+    dist_threshold: float,
+) -> Tuple[np.ndarray, np.ndarray]:
+    """Compute inlier correspondences using the ground truth triangular surface mesh of the scene. First, rays are
+    back-projected at each keypoint in the images and intersections between these rays and the ground truth mesh are
+    recorded. Next, given a match, the the mesh intersections corresponding to each keypoint is forward-projected into
```

Comment:
nit: typo here. `the the mesh intersections`

Reply:
nit: "are forward-projected ..."

Reply:
fixed

Reply:
marking as resolved, thanks